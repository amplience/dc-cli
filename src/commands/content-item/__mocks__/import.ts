import { ImportItemBuilderOptions } from '../../../interfaces/import-item-builder-options.interface';
import { ConfigurationParameters } from '../../configure';
import { Arguments } from 'yargs';

type ReturnType = boolean | 'throw';
let expectedReturn: ReturnType = true;

export const calls: Arguments<ImportItemBuilderOptions & ConfigurationParameters>[] = [];
export const setExpectedReturn = (value: ReturnType): void => {
  expectedReturn = value;
};

export const handler = async (
  argv: Arguments<ImportItemBuilderOptions & ConfigurationParameters>
): Promise<boolean> => {
  calls.push(argv);

  if (expectedReturn == 'throw') {
    throw new Error('Forced throw in test.');
  }

  return expectedReturn;
};
