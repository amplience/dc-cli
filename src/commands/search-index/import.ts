import { EnrichedSearchIndex } from './export';

type IndexName = string;
type IndexFile = string;

export const validateNoDuplicateIndexNames = (importedIndices: {
  [filename: string]: EnrichedSearchIndex;
}): void | never => {
  const nameToFilenameMap = new Map<IndexName, IndexFile[]>(); // map: name x filenames
  for (const [filename, index] of Object.entries(importedIndices)) {
    if (index.name) {
      const otherFilenames: string[] = nameToFilenameMap.get(index.name) || [];
      if (filename) {
        nameToFilenameMap.set(index.name, [...otherFilenames, filename]);
      }
    }
  }
  const uniqueDuplicateNames: [string, IndexFile[]][] = [];
  nameToFilenameMap.forEach((filenames, name) => {
    if (filenames.length > 1) {
      uniqueDuplicateNames.push([name, filenames]);
    }
  });

  if (uniqueDuplicateNames.length > 0) {
    throw new Error(
      `Indices must have unique name values. Duplicate values found:-\n${uniqueDuplicateNames
        .map(([name, filenames]) => `  name: '${name}' in files: [${filenames.map(f => `'${f}'`).join(', ')}]`)
        .join('\n')}`
    );
  }
};
