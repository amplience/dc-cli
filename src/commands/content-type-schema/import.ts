import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import { ContentTypeSchema, DynamicContent, Hub, ValidationLevel } from 'dc-management-sdk-js';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { createStream } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { TableStream } from '../../interfaces/table.interface';
import { ImportBuilderOptions } from '../../interfaces/import-builder-options.interface';
import chalk from 'chalk';
import { createContentTypeSchema } from './create.service';
import { updateContentTypeSchema } from './update.service';
import { ImportResult, loadJsonFromDirectory, UpdateStatus } from '../../services/import.service';
import { jsonResolver } from '../../common/import/json-resolver';

export const command = 'import <dir>';

export const desc = 'Import Content Type Schemas';

export interface SchemaOptions {
  validation: ValidationLevel;
}

export const builder = (yargs: Argv): void => {
  yargs.positional('dir', {
    describe: 'Directory containing Content Type Schema definitions',
    type: 'string'
  });
};

export const storedSchemaMapper = (
  schema: ContentTypeSchema,
  storedSchemas: ContentTypeSchema[]
): ContentTypeSchema => {
  const found = storedSchemas.find(stored => stored.schemaId === schema.schemaId);
  const mutatedSchema = found ? { ...schema, id: found.id } : schema;

  return new ContentTypeSchema(mutatedSchema);
};

export const doCreate = async (hub: Hub, schema: ContentTypeSchema): Promise<ContentTypeSchema> => {
  try {
    const createdSchemaType = await createContentTypeSchema(
      schema.body || '',
      schema.validationLevel || ValidationLevel.CONTENT_TYPE,
      hub
    );

    return createdSchemaType;
  } catch (err) {
    throw new Error(`Error registering content type schema with body: ${schema.body}\n\n${err}`);
  }
};

const equals = (a: ContentTypeSchema, b: ContentTypeSchema): boolean =>
  a.id === b.id && a.schemaId === b.schemaId && a.body === b.body;

export const doUpdate = async (
  client: DynamicContent,
  schema: ContentTypeSchema
): Promise<{ contentTypeSchema: ContentTypeSchema; updateStatus: UpdateStatus }> => {
  try {
    const retrievedSchema = await client.contentTypeSchemas.get(schema.id || '');
    if (equals(retrievedSchema, schema)) {
      return { contentTypeSchema: retrievedSchema, updateStatus: UpdateStatus.SKIPPED };
    }
    const updatedSchema = await updateContentTypeSchema(
      retrievedSchema,
      schema.body || '',
      schema.validationLevel || ValidationLevel.CONTENT_TYPE
    );

    return { contentTypeSchema: updatedSchema, updateStatus: UpdateStatus.UPDATED };
  } catch (err) {
    throw new Error(`Error updating content type schema ${schema.schemaId || '<unknown>'}: ${err.message}`);
  }
};

export const processSchemas = async (
  schemasToProcess: ContentTypeSchema[],
  client: DynamicContent,
  hub: Hub
): Promise<void> => {
  const tableStream = (createStream(streamTableOptions) as unknown) as TableStream;

  tableStream.write([chalk.bold('ID'), chalk.bold('Schema ID'), chalk.bold('Result')]);
  for (const schema of schemasToProcess) {
    let schemaId = schema.id;
    let status: ImportResult;
    if (schema.id) {
      const result = await doUpdate(client, schema);
      status = result.updateStatus === UpdateStatus.SKIPPED ? 'UP-TO-DATE' : 'UPDATED';
    } else {
      const result = await doCreate(hub, schema);
      schemaId = result.id || 'UNKNOWN';
      status = 'CREATED';
    }
    tableStream.write([schemaId || '', schema.schemaId || '', status]);
  }
  process.stdout.write('\n');
};

export const resolveSchemaBody = async (schemas: ContentTypeSchema[], dir: string): Promise<ContentTypeSchema[]> => {
  for (const schema of schemas) {
    if (schema.body) {
      schema.body = await jsonResolver(schema.body, dir);
    }
  }
  return schemas;
};

export const handler = async (argv: Arguments<ImportBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir } = argv;
  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const schemas = Object.values(loadJsonFromDirectory<ContentTypeSchema>(dir, ContentTypeSchema));
  await resolveSchemaBody(schemas, dir);
  const storedSchemas = await paginator(hub.related.contentTypeSchema.list);
  const schemasToProcess = schemas.map(schemas => storedSchemaMapper(schemas, storedSchemas));

  await processSchemas(schemasToProcess, client, hub);
};
