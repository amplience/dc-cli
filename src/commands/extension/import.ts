import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import { DynamicContent, Extension, Hub, ValidationLevel } from 'dc-management-sdk-js';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { table } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { ImportBuilderOptions } from '../../interfaces/import-builder-options.interface';
import chalk from 'chalk';
import { ImportResult, loadJsonFromDirectory, UpdateStatus } from '../../services/import.service';
import { FileLog } from '../../common/file-log';
import { createLog, getDefaultLogPath } from '../../common/log-helpers';
import { equals } from './export';

export const command = 'import <dir>';

export const desc = 'Import Extensions';

export interface SchemaOptions {
  validation: ValidationLevel;
}

export const LOG_FILENAME = (platform: string = process.platform): string =>
  getDefaultLogPath('extension', 'import', platform);

export const builder = (yargs: Argv): void => {
  yargs.positional('dir', {
    describe: 'Directory containing Extensions',
    type: 'string'
  });

  yargs.option('logFile', {
    type: 'string',
    default: LOG_FILENAME,
    describe: 'Path to a log file to write to.',
    coerce: createLog
  });
};

type ExtensionName = string;
type ExtensionFile = string;

export const validateNoDuplicateExtensionNames = (importedExtensions: {
  [filename: string]: Extension;
}): void | never => {
  const nameToFilenameMap = new Map<ExtensionName, ExtensionFile[]>(); // map: name x filenames
  for (const [filename, extension] of Object.entries(importedExtensions)) {
    if (extension.name) {
      const otherFilenames: string[] = nameToFilenameMap.get(extension.name) || [];
      if (filename) {
        nameToFilenameMap.set(extension.name, [...otherFilenames, filename]);
      }
    }
  }
  const uniqueDuplicateNames: [string, ExtensionFile[]][] = [];
  nameToFilenameMap.forEach((filenames, name) => {
    if (filenames.length > 1) {
      uniqueDuplicateNames.push([name, filenames]);
    }
  });

  if (uniqueDuplicateNames.length > 0) {
    throw new Error(
      `Extensions must have unique name values. Duplicate values found:-\n${uniqueDuplicateNames
        .map(([name, filenames]) => `  name: '${name}' in files: [${filenames.map(f => `'${f}'`).join(', ')}]`)
        .join('\n')}`
    );
  }
};

export const filterExtensionsById = (
  idFilter: string[],
  importedExtensions: {
    [filename: string]: Extension;
  }
): void | never => {
  for (const [filename, extension] of Object.entries(importedExtensions)) {
    if (idFilter.indexOf(extension.id as string) === -1) {
      delete importedExtensions[filename];
    }
  }
};

export const storedExtensionMapper = (extension: Extension, storedExtensions: Extension[]): Extension => {
  const found = storedExtensions.find(stored => stored.name === extension.name);
  const mutatedExtension = found ? { ...extension, id: found.id } : extension;

  return new Extension(mutatedExtension);
};

export const doCreate = async (hub: Hub, extension: Extension, log: FileLog): Promise<Extension> => {
  try {
    const createdExtension = await hub.related.extensions.create(extension);

    log.addAction('CREATE', `${createdExtension.id}`);

    return createdExtension;
  } catch (err) {
    throw new Error(`Error creating extension ${extension.name}:\n\n${err}`);
  }
};

export const doUpdate = async (
  client: DynamicContent,
  extension: Extension,
  log: FileLog
): Promise<{ extension: Extension; updateStatus: UpdateStatus }> => {
  try {
    const retrievedExtension: Extension = await client.extensions.get(extension.id || '');
    if (equals(retrievedExtension, extension)) {
      return { extension: retrievedExtension, updateStatus: UpdateStatus.SKIPPED };
    }

    const updatedExtension = await retrievedExtension.related.update(extension);

    log.addAction('UPDATE', `${retrievedExtension.id}`);

    return { extension: updatedExtension, updateStatus: UpdateStatus.UPDATED };
  } catch (err) {
    throw new Error(`Error updating extension ${extension.name}: ${err.message}`);
  }
};

export const processExtensions = async (
  extensionsToProcess: Extension[],
  client: DynamicContent,
  hub: Hub,
  log: FileLog
): Promise<void> => {
  const data: string[][] = [];

  data.push([chalk.bold('ID'), chalk.bold('Name'), chalk.bold('Result')]);
  for (const entry of extensionsToProcess) {
    let status: ImportResult;
    let extension: Extension;
    if (entry.id) {
      const result = await doUpdate(client, entry, log);
      extension = result.extension;
      status = result.updateStatus === UpdateStatus.SKIPPED ? 'UP-TO-DATE' : 'UPDATED';
    } else {
      extension = await doCreate(hub, entry, log);
      status = 'CREATED';
    }
    data.push([extension.id || '', extension.name, status]);
  }

  log.appendLine(table(data, streamTableOptions));
};

export const handler = async (
  argv: Arguments<ImportBuilderOptions & ConfigurationParameters>,
  idFilter?: string[]
): Promise<void> => {
  const { dir, logFile } = argv;
  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const log = logFile.open();
  const extensions = loadJsonFromDirectory<Extension>(dir, Extension);
  if (Object.keys(extensions).length === 0) {
    throw new Error(`No extensions found in ${dir}`);
  }

  validateNoDuplicateExtensionNames(extensions);

  if (idFilter) {
    filterExtensionsById(idFilter, extensions);
  }

  const storedExtensions = await paginator(hub.related.extensions.list);
  const extensionsToProcess = Object.values(extensions).map(extension =>
    storedExtensionMapper(extension, storedExtensions)
  );

  await processExtensions(extensionsToProcess, client, hub, log);

  await log.close();
};
