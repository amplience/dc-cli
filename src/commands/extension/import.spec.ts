import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import { Extension, Hub } from 'dc-management-sdk-js';
import * as importModule from './import';
import {
  builder,
  command,
  doCreate,
  doUpdate,
  handler,
  LOG_FILENAME,
  processExtensions,
  storedExtensionMapper,
  validateNoDuplicateExtensionNames
} from './import';
import Yargs from 'yargs/yargs';
import { table } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { loadJsonFromDirectory, UpdateStatus } from '../../services/import.service';
import chalk from 'chalk';
import { FileLog } from '../../common/file-log';
import { createLog, getDefaultLogPath } from '../../common/log-helpers';

jest.mock('../../services/dynamic-content-client-factory');
jest.mock('../../view/data-presenter');
jest.mock('../../services/import.service');
jest.mock('../../common/dc-management-sdk-js/paginator');
jest.mock('fs');
jest.mock('table');
jest.mock('../../common/log-helpers');

describe('extension import command', (): void => {
  afterEach((): void => {
    jest.restoreAllMocks();
  });

  it('should implement an import command', () => {
    expect(command).toEqual('import <dir>');
  });

  describe('builder tests', () => {
    it('should configure yargs', () => {
      const argv = Yargs(process.argv.slice(2));
      const spyPositional = jest.spyOn(argv, 'positional').mockReturnThis();
      const spyOption = jest.spyOn(argv, 'option').mockReturnThis();

      builder(argv);

      expect(spyPositional).toHaveBeenCalledWith('dir', {
        describe: 'Directory containing Extensions',
        type: 'string'
      });

      expect(spyOption).toHaveBeenCalledWith('logFile', {
        type: 'string',
        default: LOG_FILENAME,
        describe: 'Path to a log file to write to.',
        coerce: createLog
      });
    });
  });

  describe('storedExtensionMapper', () => {
    it('it should map to a stored extension', () => {
      const importedExtension = new Extension({
        name: 'matched-name',
        label: 'mutated-label'
      });
      const storedExtension = [new Extension({ id: 'stored-id', name: 'matched-name', label: 'label' })];
      const result = storedExtensionMapper(importedExtension, storedExtension);

      expect(result).toEqual(
        expect.objectContaining({
          id: 'stored-id',
          name: 'matched-name',
          label: 'mutated-label'
        })
      );
    });

    it('should not map to a stored extension', () => {
      const importedExtension = new Extension({
        name: 'not-matched-name',
        label: 'mutated-label'
      });
      const storedExtension = [new Extension({ id: 'stored-id', name: 'matched-name', label: 'label' })];
      const result = storedExtensionMapper(importedExtension, storedExtension);

      expect(result).toEqual(expect.objectContaining({ name: 'not-matched-name', label: 'mutated-label' }));
    });
  });

  describe('doCreate', () => {
    it('should create a extension and return report', async () => {
      const mockHub = new Hub();
      const log = new FileLog();
      const newExtension = new Extension({ id: 'created-id' });
      const mockCreate = jest.fn().mockResolvedValue(newExtension);
      mockHub.related.extensions.create = mockCreate;
      const extension = { name: 'extension-name', label: 'test-label' };
      const result = await doCreate(mockHub, extension as Extension, log);

      expect(log.getData('CREATE')).toMatchInlineSnapshot(`
        Array [
          "created-id",
        ]
      `);
      expect(mockCreate).toHaveBeenCalledWith(expect.objectContaining(extension));
      expect(result).toEqual(newExtension);
    });

    it('should throw an error when extension create fails', async () => {
      const mockHub = new Hub();
      const log = new FileLog();
      const mockCreate = jest.fn().mockImplementation(() => {
        throw new Error('Error creating extension');
      });
      mockHub.related.extensions.create = mockCreate;
      const extension = { name: 'extension-name', label: 'test-label' };

      await expect(doCreate(mockHub, extension as Extension, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('UPDATE')).toEqual([]);
    });

    it('should throw an error when extension create fails if a string error is returned by the sdk', async () => {
      const mockHub = new Hub();
      const log = new FileLog();
      const mockCreate = jest
        .fn()
        .mockRejectedValue(
          'The create-extension action is not available, ensure you have permission to perform this action.'
        );
      mockHub.related.extensions.create = mockCreate;
      const extension = { name: 'extension-name', label: 'test-label' };

      await expect(doCreate(mockHub, extension as Extension, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('UPDATE')).toEqual([]);
    });
  });

  describe('doUpdate', () => {
    const mockGet = jest.fn();
    let mockDynamicContentClientFactory: jest.Mock;

    beforeEach(() => {
      mockDynamicContentClientFactory = (dynamicContentClientFactory as jest.Mock).mockReturnValue({
        extensions: {
          get: mockGet
        }
      });
    });
    it('should update a extension and return report', async () => {
      const mutatedExtension = {
        id: 'stored-id',
        name: 'not-matched-name',
        label: 'mutated-label'
      } as Extension;
      const storedExtension = new Extension({
        id: 'stored-id',
        name: 'matched-name',
        label: 'label'
      });
      const expectedExtension = new Extension({
        id: 'stored-id',
        name: 'not-matched-name',
        label: 'mutated-label'
      });
      mockGet.mockResolvedValue(storedExtension);

      const updatedExtension = new Extension(mutatedExtension);
      const mockUpdate = jest.fn().mockResolvedValue(updatedExtension);
      storedExtension.related.update = mockUpdate;
      const client = mockDynamicContentClientFactory();
      const log = new FileLog();
      const result = await doUpdate(client, mutatedExtension, log);

      expect(log.getData('UPDATE')).toMatchInlineSnapshot(`
        Array [
          "stored-id",
        ]
      `);
      expect(result).toEqual({ extension: updatedExtension, updateStatus: UpdateStatus.UPDATED });
      expect(mockUpdate).toHaveBeenCalledWith(expectedExtension.toJSON());
    });

    it('should skip update when no change to extension and return report', async () => {
      const mutatedExtension = new Extension({
        id: 'stored-id',
        name: 'matched-name',
        label: 'label'
      });
      const storedExtension = new Extension({
        id: 'stored-id',
        name: 'matched-name',
        label: 'label'
      });
      mockGet.mockResolvedValue(storedExtension);
      const client = mockDynamicContentClientFactory();
      const log = new FileLog();
      const result = await doUpdate(client, mutatedExtension, log);

      expect(result).toEqual({ extension: storedExtension, updateStatus: UpdateStatus.SKIPPED });
      expect(log.getData('UPDATE')).toEqual([]);
    });

    it('should throw an error when unable to get extension during update', async () => {
      const mutatedExtension = {
        id: 'stored-id',
        name: 'matched-name',
        label: 'label'
      } as Extension;
      mockGet.mockImplementation(() => {
        throw new Error('Error retrieving extension');
      });
      const client = mockDynamicContentClientFactory();
      const log = new FileLog();

      await expect(doUpdate(client, mutatedExtension, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('UPDATE')).toEqual([]);
    });

    it('should throw an error when unable to update extension during update if a string error is returned by sdk', async () => {
      const mutatedExtension = new Extension({
        id: 'stored-id',
        name: 'not-matched-name',
        label: 'mutated-label'
      });
      const storedExtension = new Extension({
        id: 'stored-id',
        name: 'matched-name',
        label: 'label'
      });
      const mockUpdate = jest
        .fn()
        .mockRejectedValue('The update action is not available, ensure you have permission to perform this action.');
      storedExtension.related.update = mockUpdate;
      mockGet.mockResolvedValue(storedExtension);
      const client = mockDynamicContentClientFactory();
      const log = new FileLog();
      await expect(doUpdate(client, mutatedExtension, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('UPDATE')).toEqual([]);
      expect(mockUpdate).toHaveBeenCalledWith(mutatedExtension);
    });

    it('should throw an error when unable to update extension during update', async () => {
      const mutatedExtension = new Extension({
        id: 'stored-id',
        name: 'not-matched-name',
        label: 'mutated-label'
      });
      const storedExtension = new Extension({
        id: 'stored-id',
        name: 'matched-name',
        label: 'label'
      });
      const mockUpdate = jest.fn().mockRejectedValue(new Error('Error saving extension'));
      storedExtension.related.update = mockUpdate;
      mockGet.mockResolvedValue(storedExtension);
      const client = mockDynamicContentClientFactory();
      const log = new FileLog();
      await expect(doUpdate(client, mutatedExtension, log)).rejects.toThrowErrorMatchingSnapshot();
      expect(log.getData('UPDATE')).toEqual([]);
      expect(mockUpdate).toHaveBeenCalledWith(mutatedExtension);
    });
  });

  describe('processExtensions', () => {
    let mockTable: jest.Mock;

    beforeEach(() => {
      mockTable = table as jest.Mock;
      mockTable.mockImplementation(jest.requireActual('table').table);
    });

    it('should create and update a extension', async () => {
      const client = (dynamicContentClientFactory as jest.Mock)();
      const hub = new Hub();
      const extensionsToProcess = [
        new Extension({
          name: 'extension-name',
          label: 'created'
        }),
        new Extension({
          id: 'updated-id',
          name: 'extension-name-2',
          label: 'updated'
        }),
        new Extension({
          id: 'up-to-date-id',
          name: 'extension-name-3',
          label: 'up-to date'
        })
      ];

      const createdExtension = new Extension({
        id: 'created-id',
        ...extensionsToProcess[0].toJSON()
      });
      jest.spyOn(importModule, 'doCreate').mockResolvedValueOnce(createdExtension);
      const doUpdateResult1 = {
        extension: extensionsToProcess[1],
        updateStatus: UpdateStatus.UPDATED
      };
      jest.spyOn(importModule, 'doUpdate').mockResolvedValueOnce(doUpdateResult1);
      const doUpdateResult2 = {
        extension: extensionsToProcess[2],
        updateStatus: UpdateStatus.SKIPPED
      };
      jest.spyOn(importModule, 'doUpdate').mockResolvedValueOnce(doUpdateResult2);

      await processExtensions(extensionsToProcess, client, hub, new FileLog());

      expect(importModule.doCreate).toHaveBeenCalledWith(hub, extensionsToProcess[0], expect.any(FileLog));
      expect(importModule.doUpdate).toHaveBeenCalledWith(client, extensionsToProcess[1], expect.any(FileLog));

      expect(mockTable).toHaveBeenCalledTimes(1);
      expect(mockTable).toHaveBeenNthCalledWith(
        1,
        [
          [chalk.bold('ID'), chalk.bold('Name'), chalk.bold('Result')],
          [createdExtension.id, createdExtension.name, 'CREATED'],
          [doUpdateResult1.extension.id, doUpdateResult1.extension.name, 'UPDATED'],
          [doUpdateResult2.extension.id, doUpdateResult2.extension.name, 'UP-TO-DATE']
        ],
        streamTableOptions
      );
    });
  });

  describe('validateNoDuplicateExtensionNames', function() {
    it('should not throw an error when there are no duplicates', () => {
      const extensionsToProcess = {
        'file-1': new Extension({
          name: 'extension-name-1'
        }),
        'file-2': new Extension({
          name: 'extension-name-2'
        })
      };

      expect(() => validateNoDuplicateExtensionNames(extensionsToProcess)).not.toThrow();
    });

    it('should throw and error when there are duplicate uris', () => {
      const extensionsToProcess = {
        'file-1': new Extension({
          name: 'extension-name-1'
        }),
        'file-2': new Extension({
          name: 'extension-name-2'
        }),
        'file-3': new Extension({
          name: 'extension-name-2'
        }),
        'file-4': new Extension({
          name: 'extension-name-1'
        })
      };

      expect(() => validateNoDuplicateExtensionNames(extensionsToProcess)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('filterExtensionsById', function() {
    it('should delete extensions without a matching id', () => {
      const extensionsToProcess = {
        'file-1': new Extension({
          id: 'extension-id-1',
          name: 'extension-name-1'
        }),
        'file-2': new Extension({
          id: 'extension-id-2',
          name: 'extension-name-2'
        })
      };

      const expectedResult = { 'file-2': extensionsToProcess['file-2'] };

      importModule.filterExtensionsById(['extension-id-2'], extensionsToProcess);

      expect(extensionsToProcess).toEqual(expectedResult);
    });

    it('should remove all extensions if no ids are given', () => {
      const extensionsToProcess = {
        'file-1': new Extension({
          id: 'extension-id-1',
          name: 'extension-name-1'
        }),
        'file-2': new Extension({
          id: 'extension-id-2',
          name: 'extension-name-2'
        })
      };

      importModule.filterExtensionsById([], extensionsToProcess);

      expect(extensionsToProcess).toEqual({});
    });
  });

  describe('handler tests', () => {
    const yargArgs = {
      $0: 'test',
      _: ['test']
    };
    const config = {
      clientId: 'client-id',
      clientSecret: 'client-id',
      hubId: 'hub-id'
    };

    const mockGetHub = jest.fn();

    beforeEach(() => {
      (dynamicContentClientFactory as jest.Mock).mockReturnValue({
        hubs: {
          get: mockGetHub
        }
      });

      mockGetHub.mockResolvedValue({
        related: {
          extensions: {
            list: jest.fn()
          }
        }
      });
    });

    it('should use getDefaultLogPath for LOG_FILENAME with process.platform as default', function() {
      LOG_FILENAME();

      expect(getDefaultLogPath).toHaveBeenCalledWith('extension', 'import', process.platform);
    });

    it('should create a extension and update', async (): Promise<void> => {
      const argv = { ...yargArgs, ...config, dir: 'my-dir', logFile: new FileLog() };
      const fileNamesAndExtensionsToImport = {
        'file-1': new Extension({
          name: 'extension-name-1',
          label: 'created'
        }),
        'file-2': new Extension({
          id: 'content-extension-id',
          name: 'extension-name-2',
          label: 'updated'
        })
      };

      (loadJsonFromDirectory as jest.Mock).mockReturnValue(fileNamesAndExtensionsToImport);
      mockGetHub.mockResolvedValue(new Hub({ id: 'hub-id' }));
      jest
        .spyOn(importModule, 'storedExtensionMapper')
        .mockReturnValueOnce(fileNamesAndExtensionsToImport['file-1'])
        .mockReturnValueOnce(fileNamesAndExtensionsToImport['file-2']);
      jest.spyOn(importModule, 'processExtensions').mockResolvedValueOnce();

      await handler(argv);

      expect(loadJsonFromDirectory).toHaveBeenCalledWith('my-dir', Extension);
      expect(mockGetHub).toHaveBeenCalledWith('hub-id');
      expect(processExtensions).toHaveBeenCalledWith(
        Object.values(fileNamesAndExtensionsToImport),
        expect.any(Object),
        expect.any(Object),
        expect.any(FileLog)
      );
    });

    it('should call filterExtensionsById when a list of ids is provided', async (): Promise<void> => {
      const argv = { ...yargArgs, ...config, dir: 'my-dir', logFile: new FileLog() };
      const fileNamesAndExtensionsToImport = {
        'file-1': new Extension({
          id: 'extension-id-1',
          name: 'extension-name-1',
          label: 'created'
        }),
        'file-2': new Extension({
          id: 'extension-id-2',
          name: 'extension-name-2',
          label: 'updated'
        })
      };

      (loadJsonFromDirectory as jest.Mock).mockReturnValue({ ...fileNamesAndExtensionsToImport });
      mockGetHub.mockResolvedValue(new Hub({ id: 'hub-id' }));
      jest.spyOn(importModule, 'storedExtensionMapper').mockReturnValueOnce(fileNamesAndExtensionsToImport['file-2']);
      jest.spyOn(importModule, 'processExtensions').mockResolvedValueOnce();

      await handler(argv, ['extension-id-2']);

      expect(loadJsonFromDirectory).toHaveBeenCalledWith('my-dir', Extension);
      expect(mockGetHub).toHaveBeenCalledWith('hub-id');
      expect(processExtensions).toHaveBeenCalledWith(
        [fileNamesAndExtensionsToImport['file-2']],
        expect.any(Object),
        expect.any(Object),
        expect.any(FileLog)
      );
    });

    it('should throw an error when no content found in import directory', async (): Promise<void> => {
      const argv = { ...yargArgs, ...config, dir: 'my-empty-dir', logFile: new FileLog() };

      (loadJsonFromDirectory as jest.Mock).mockReturnValue([]);

      await expect(handler(argv)).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
