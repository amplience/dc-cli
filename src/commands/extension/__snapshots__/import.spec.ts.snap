// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension import command doCreate should throw an error when extension create fails 1`] = `
"Error creating extension extension-name:

Error: Error creating extension"
`;

exports[`extension import command doCreate should throw an error when extension create fails if a string error is returned by the sdk 1`] = `
"Error creating extension extension-name:

The create-extension action is not available, ensure you have permission to perform this action."
`;

exports[`extension import command doUpdate should throw an error when unable to get extension during update 1`] = `"Error updating extension matched-name: Error retrieving extension"`;

exports[`extension import command doUpdate should throw an error when unable to update extension during update 1`] = `"Error updating extension not-matched-name: Error saving extension"`;

exports[`extension import command doUpdate should throw an error when unable to update extension during update if a string error is returned by sdk 1`] = `"Error updating extension not-matched-name: undefined"`;

exports[`extension import command handler tests should throw an error when no content found in import directory 1`] = `"No extensions found in my-empty-dir"`;

exports[`extension import command validateNoDuplicateExtensionNames should throw and error when there are duplicate uris 1`] = `
"Extensions must have unique name values. Duplicate values found:-
  name: 'extension-name-1' in files: ['file-1', 'file-4']
  name: 'extension-name-2' in files: ['file-2', 'file-3']"
`;
